src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -127,9 +127,54 @@ namespace winrt::TerminalApp::implementation
         _tabView.TabItemsChanged({ this, &TerminalPage::_OnTabItemsChanged });
 
         _CreateNewTabFlyout();
-        _OpenNewTab(nullptr);
 
         _tabContent.SizeChanged({ this, &TerminalPage::_OnContentSizeChanged });
+
+        // Actually start the terminal.
+        if (_appArgs.GetStartupActions().size() == 0)
+        {
+            _OpenNewTab(nullptr);
+        }
+        else
+        {
+            _appArgs.ValidateStartupCommands();
+
+            // This will kick off a chain of events to perform each startup
+            // action. As each startup action is completed, the next will be
+            // fired.
+            _processNextStartupAction();
+        }
+    }
+
+    // Method Description:
+    // - Process the next startup action in our list of startup actions. When
+    //   that action is complete, fire the next (if there are any more).
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void TerminalPage::_processNextStartupAction()
+    {
+        // If there are no actions left, do nothing.
+        if (_appArgs.GetStartupActions().size() == 0)
+        {
+            return;
+        }
+
+        // Get the next action to be processed
+        auto nextAction = _appArgs.GetStartupActions().front();
+        _appArgs.GetStartupActions().pop_front();
+
+        // Handle it on the UI thread.
+        Dispatcher().RunAsync(CoreDispatcherPriority::Low, [weakThis{ get_weak() }, nextAction]() {
```

Comment:
maybe normal priority for this?

Reply:
and we're doing this on a bunch of different dispatch queues so as to not starve the UI thread during startup?
is it unsafe to just `for(actions) { do them }` on the UI thread?